xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEBuffer.cpp

```diff
@@ -220,21 +243,31 @@ bool CActiveAEBufferPoolResample::Create(unsigned int totaltime, bool remap, boo
    */
   if ((useDSP || m_changeDSP) && !m_bypassDSP)
   {
-    m_dspFormat = m_inputFormat;
-    m_useDSP = CServiceBroker::GetADSP().CreateDSPs(m_streamId, m_processor, m_dspFormat, m_format, upmix, m_resampleQuality, m_MatrixEncoding, m_AudioServiceType, m_Profile);
+    // use input format from this CActiveAEBufferPoolResample as input for AudioDSP
+    // and use CActiveAEBufferPool::m_format as output format of AudioDSP
+    m_useDSP = CServiceBroker::GetADSP().CreateDSPs(m_streamId, m_processor, m_inputFormat, CActiveAEBufferPool::m_format, 
+                                                    upmix, m_resampleQuality, m_MatrixEncoding, m_AudioServiceType, m_Profile);
     if (m_useDSP)
     {
-
-      m_inputFormat.m_channelLayout = m_processor->GetChannelLayout();    /* Overide input format with DSP's supported format */
-      m_inputFormat.m_sampleRate    = m_processor->GetOutputSamplerate(); /* Overide input format with DSP's generated samplerate */
-      m_inputFormat.m_dataFormat    = m_processor->GetDataFormat();       /* Overide input format with DSP's processed data format, normally it is float */
-      m_inputFormat.m_frames        = m_processor->GetOutputFrames();
+      m_adspOutFormat = CActiveAEBufferPool::m_format;
+      m_adspOutFormat.m_channelLayout = m_processor->GetChannelLayout();    /* Overide input format with DSP's supported format */
+      m_adspOutFormat.m_sampleRate    = m_processor->GetOutputSamplerate(); /* Overide input format with DSP's generated samplerate */
+      m_adspOutFormat.m_dataFormat    = m_processor->GetDataFormat();       /* Overide input format with DSP's processed data format, normally it is float */
+      m_adspOutFormat.m_frames        = m_processor->GetOutputFrames();
       m_forceResampler              = true;                               /* Force load of ffmpeg resampler, required to detect exact input and output channel alignment pointers */
       if (m_processor->GetChannelLayout().Count() > 2)                    /* Disable upmix for CActiveAEResample if DSP layout > 2.0, becomes perfomed by DSP */
         upmix = false;
 
-      m_dspBuffer = new CActiveAEBufferPool(m_inputFormat);               /* Get dsp processing buffer class, based on dsp output format */
+      if (!m_dspBuffer)
+      {
+        m_dspBuffer = new CActiveAEBufferPool(m_adspOutFormat); /* Get dsp processing buffer class, based on dsp output format */
+      }
+      else
+      {
+        m_dspBuffer->m_format = m_adspOutFormat;
+      }
       m_dspBuffer->Create(totaltime);
+      CActiveAEBufferPool::Create(totaltime); // recreate output buffer cause format has changed
```

Comment:
Think in objects, not in functions. Having adsp in CActiveAEBufferPool is right because it manipulates the buffers like resampler does. In a clean design adsp would have its own stage, hence its own CActiveAEBufferPool where it operates on.

But again, a buffer pool is not supposed to recreate itself. This needs to be changed.